test/integration/targets/git/tasks/separate-git-dir.yml

```diff
@@ -0,0 +1,81 @@
+# test code for repositories with separate git dir updating
+# see https://github.com/ansible/ansible/pull/38016
+# see https://github.com/ansible/ansible/issues/30034
+
+- name: SEPARATE-GIT-DIR | clear checkout_dir
+  file:
+    state: absent
+    path: '{{ checkout_dir }}'
+
+- name: create a tempdir for separate git dir
+  local_action: shell mktemp -du
+  register: tempdir
+
+- name: SEPARATE-GIT-DIR | clone with a separate git dir
+  command: git clone {{ repo_format1 }} {{ checkout_dir }} --separate-git-dir={{ tempdir.stdout }}
+
+- name: SEPARATE-GIT-DIR | update repo the usual way
+  git:
+    repo: "{{ repo_format1 }}"
+    dest: "{{ checkout_dir }}"
+
+- name: SEPARATE-GIT-DIR | set git dir to non-existent dir
+  shell: "echo gitdir: /dev/null/non-existent-dir > .git"
+  args:
+    chdir: "{{ checkout_dir }}"
+
+- name: SEPARATE-GIT-DIR | update repo the usual way
+  git:
+    repo: "{{ repo_format1 }}"
+    dest: "{{ checkout_dir }}"
+  ignore_errors: yes
+  register: result
+
+- name: SEPARATE-GIT-DIR | check update has failed
+  assert:
+    that:
+      - result is failed
+
+- name: SEPARATE-GIT-DIR | set .git file to bad format
+  shell: "echo some text  gitdir: {{ checkout_dir }} > .git"
+  args:
+    chdir: "{{ checkout_dir }}"
+
+- name: SEPARATE-GIT-DIR | update repo the usual way
+  git:
+    repo: "{{ repo_format1 }}"
+    dest: "{{ checkout_dir }}"
+  ignore_errors: yes
+  register: result
+
+- name: SEPARATE-GIT-DIR | check update has failed
+  assert:
+    that:
+      - result is failed
+
+- name: SEPARATE-GIT-DIR | set .git file to bad format, when gitdir is on the second line
+  shell: "echo \\ngitdir: {{ checkout_dir }} > .git"
+  args:
+    chdir: "{{ checkout_dir }}"
+
+- name: SEPARATE-GIT-DIR | update repo the usual way
+  git:
+    repo: "{{ repo_format1 }}"
+    dest: "{{ checkout_dir }}"
+  ignore_errors: yes
+  register: result
+
+- name: SEPARATE-GIT-DIR | check update has failed
+  assert:
+    that:
+      - result is failed
```

Comment:
it should not fails if `.gitdir` contains a list of `key: value`

Reply:
I can't find any reference that proves that. Git documentation [says](https://git-scm.com/docs/gitrepository-layout#_description):
>  Also you can have a plain text file .git at the root of your working tree, containing gitdir: <path> to point at the real directory that has the repository.

And no information about the key-value list.

I also tried a few examples, including
```
key: value
gitdir: ...path
```
and (with newline at the front):
```

gitdir: ...path
```

And all of the examples fail at the repository update stage with the message from git: `fatal: Invalid gitfile format: .git`. Can you please provide any correct examples of key-value lists for the `.git` file which will work with Git?

Reply:
Not really, but since that other snippet loops through key-value lines there must be something working this way. I'll try investigating this.

Reply:
It looks like the best way of checking integrity of the repo is [`git-fsck`](https://git-scm.com/docs/git-fsck), but I don't thing that it's needed here now.

Reply:
But you are right, it does contain a validation for a file immediately starting with `gitdir: `: https://github.com/git/git/blob/master/setup.c#L592

Reply:
So now we know that it should only contain `gitdir: ...path` and nothing more in that file. Trailing EOLs are being stripped off by git.